File: week35/CoinJam/Main.java
package mx.letmethink;

import java.util.List;
import java.io.*;
import java.util.*;
import java.math.BigInteger;

public class Main {
    public static void main(String[] args) {
        InputStream inputStream = System.in;
        OutputStream outputStream = System.out;
        InputReader in = new InputReader(inputStream);
        PrintWriter out = new PrintWriter(outputStream);
        Solver solver = new Solver();
        solver.solve(in, out);
        out.close();
    }
}


class Solution {
    private final String coin;
    private final List<BigInteger> divisors;

    public Solution(long mask, final List<BigInteger> divisors, int bitSize) {
        char[] bits = new char[bitSize];
        for (int i = 0; i < bitSize; i++) {
            if ((mask & (1L << i)) != 0) {
                bits[i] = '1';
            } else {
                bits[i] = '0';
            }
        }
        this.coin = new String(bits);
        this.divisors = divisors;
    }

    public String getCoin() {
        return coin;
    }

    public final List<BigInteger> getDivisors() {
        return divisors;
    }

}

/** Built using CHelper plug-in, Actual solution is at the top. */
class Solver {
    private static final int MAX = 1 << 22;
    private long[] sieve;
    private List<BigInteger> primes;

    public void solve(InputReader in, PrintWriter out) {
        init();
        int T = in.nextInt();
        for (int tc = 1; tc <= T; tc++) {
            int N = in.nextInt();
            int J = in.nextInt();
            System.out.println("Case #" + tc + ":");
            for (Solution solution : findSolutions(N, J)) {
                System.out.print(solution.getCoin());
                for (BigInteger divisor : solution.getDivisors()) {
                    System.out.print(" " + divisor);
                }
                System.out.println();
            }
        }
    }

    void init() {
        sieve = new long[MAX];
        primes = new ArrayList<>();

        long sr = (long)Math.sqrt(MAX);
        for (int i = 2; i <= sr; i++) {
            if (sieve[i] != 0) {
                continue;
            }
            for (int j = i * i; j < MAX; j += i) {
                sieve[j] = i;
            }
        }
        for (long i = 2; i < MAX; i++) {
            if (sieve[(int)i] == 0) {
                primes.add(new BigInteger(i + ""));
            }
            if (primes.size() == 5000) {
                break;
            }
        }
    }

    BigInteger convert(long mask, int base, int bitSize) {
        char[] digits = new char[bitSize];
        for (int i = 0; i < bitSize; i++) {
            digits[i] = ((mask & (1L << i)) != 0) ? '1' : '0';
        }
        return new BigInteger(new String(digits), base);
    }


    public List<Solution> findSolutions(int N, int J) {
        List<Solution> solutions = new ArrayList<>();
        for (long mask = 0; mask < (1L << (N - 2)); mask++) {
            long candidate = (mask << 1) | 1L | (1L << (N - 1));
            List<BigInteger> divisors = new ArrayList<>();
            for (int base = 2; base <= 10; base++) {
                BigInteger x = convert(candidate, base, N);
                BigInteger divisor = null;
                for (BigInteger p : primes) {
                    if ((p.compareTo(x) < 0) && (x.mod(p).equals(BigInteger.ZERO))) {
                        divisor = p;
                        break;
                    }
                }
                if (divisor != null) {
                    divisors.add(divisor);
                }
            }
            if (divisors.size() == 9) {
                solutions.add(new Solution(candidate, divisors, N));
            }
            if (solutions.size() == J) {
                break;
            }
        }
        return solutions;
    }


}

//region
class InputReader {
    private BufferedReader reader;
    private StringTokenizer tokenizer;

    public InputReader(InputStream stream)
    {
        reader = new BufferedReader(new InputStreamReader(stream));
        tokenizer = null;
    }

    public String next()
    {
        while (tokenizer == null || !tokenizer.hasMoreTokens()) {
            try {
                tokenizer = new StringTokenizer(reader.readLine());
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        return tokenizer.nextToken();
    }

    public int nextInt()    { return Integer.parseInt(next());  }
    public long nextLong()  { return Long.parseLong(next());    }
}

File: week35/CoinJam/main.cpp
/* Copyright 2017 Rafael Rendón Pablo <rafaelrendonpablo@gmail.com> */
// region Template
#include <iostream>
#include <iomanip>
#include <fstream>
#include <sstream>
#include <string>
#include <bitset>
#include <vector>
#include <queue>
#include <map>
#include <set>
#include <cmath>
#include <cctype>
#include <cassert>
#include <cstring>
#include <numeric>
#include <iomanip>
#include <algorithm>

using std::string;
using std::vector;
using std::map;
using std::set;
using std::queue;

typedef long long           int64;
typedef unsigned long long  uint64;
const double kEps   = 10e-8;
const int kMax      = 1 << 22;
const int kInf      = 1 << 30;
// endregion

int sieve[kMax];
vector<int64> primes;

class Solution {
public:
    Solution() { }

    Solution(int64 mask, const vector<int64>& divisors, int bitsize) {
        for (int i = 0; i < bitsize; i++) {
            if (mask & (1LL << i)) {
                coin += "1";
            } else {
                coin += "0";
            }
        }
        this->divisors = divisors;
    }

    const string& getCoin() {
        return coin;
    }

    const vector<int64>& getDivisors() {
        return divisors;
    }

private:
    string coin;
    vector<int64> divisors;
};

int64 convert(int64 mask, int base, int size) {
    int64 v = 0;
    for (int i = 0; i < size; i++) {
        int digit = ((mask & (1LL << i)) != 0) ? 1 : 0;
        v = v * base + digit;
    }
    return v;
}

void init() {
    int sr = std::sqrt(kMax);
    for (int i = 2; i <= sr; i++) {
        if (sieve[i] != 0) {
            continue;
        }
        for (int j = i * i; j < kMax; j += i) {
            sieve[j] = i;
        }
    }
    for (int i = 2; i < kMax; i++) {
        if (sieve[i] == 0) {
            primes.push_back(i);
        }
        if (primes.size() == 5000) {
            break;
        }
    }
}

vector<Solution> findSolutions(int N, int J) {
    vector<Solution> solutions;
    for (int64 mask = 0; mask < (1LL << (N - 2)); mask++) {
        int64 candidate = (mask << 1) | 1LL | (1LL << (N - 1));
        vector<int64> divisors;
        for (int base = 2; base <= 10; base++) {
            int64 x = convert(candidate, base, N);
            int64 divisor = -1;
            for (int64 p : primes) {
                if ((p < x) && (x % p == 0)) {
                    divisor = p;
                    break;
                }
            }
            if (x == 1590504037202545653841921LL) {
                std::cout << "div = " << divisor << std::endl;
            }
            if (divisor != -1) {
                divisors.push_back(divisor);
            }
        }
        if (divisors.size() == 9) {
            solutions.push_back(Solution(candidate, divisors, N));
        }
        if (int(solutions.size()) == J) {
            break;
        }
    }
    return solutions;
}

int main() {
    init();
    int T;
    std::cin >> T;
    for (int tc = 1; tc <= T; tc++) {
        int N, J;
        std::cin >> N >> J;
        std::cout << "Case #" << tc << ":\n";

        for (Solution solution : findSolutions(N, J)) {
            std::cout << solution.getCoin() << " ";
            for (int divisor : solution.getDivisors()) {
                std::cout << " " << divisor;
            }
            std::cout << std::endl;
        }
    }
    return EXIT_SUCCESS;
}
